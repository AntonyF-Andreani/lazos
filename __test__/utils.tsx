import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";

import theme from "theme";

/**
 * Colocar todos los providers, se puede mockear redux y las funciones
 */
const AllTheProviders: React.FC = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui: React.ReactElement) => render(ui, { wrapper: AllTheProviders });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
