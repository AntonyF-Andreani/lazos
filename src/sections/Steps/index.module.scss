.section {
  margin-top: 110px;
  margin-bottom: 20px;
}

.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  text-align: center;

  @media only screen and (min-width: 1000px) {
    background: transparent url("../../assets/images/miscellaneous-steps.png") no-repeat center 150px;
  }
}

.imageContainer {
  position: absolute;
  margin: auto;
  height: 349;
  width: 1348;
}

.stepsContainer {
  display: flex;
  flex-flow: row wrap;
  align-items: flex-start;
  justify-content: center;
  margin-top: 50px;
  row-gap: 40px;

  @media only screen and (min-width: 1000px) {
    flex-wrap: nowrap;
  }
}

.sectionTitle {
  padding-inline: 15px;
  padding-bottom: 40px;
  color: #212529;
  font-size: 2.3em;
  font-weight: 700;

  @media (max-width: 425px) {
    font-size: 2em;
  }
}

.step {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 160px;
  max-width: 160px;
  margin-inline: 15px;

  @media only screen and (min-width: 1000px) {
    &[data-key="0"] {
      margin-top: 100px;
    }

    &[data-key="1"] {
      margin-top: 80px;
    }

    &[data-key="2"] {
      margin-top: 100px;
    }

    &[data-key="3"] {
      max-width: 400px;
      width: 400px;
      margin-left: 30px;
    }
  }
}

.stepsImageContainer {
  border-radius: 150px;
  width: 145px;
  height: 145px;
}

.stepImage {
  border-radius: 150px;
  margin: 7px 8px;

  // box-shadow: 0px 0px 0px 5px #ffffff,
  // 0px 0px 0px 7px #c7c7c7;

  @media only screen and (min-width: 1000px) {
    &[data-key="3"] {
      max-width: 338px;
      max-height: 338px;
      width: 338px;
      height: 338px;
      border-radius: 300px;
    }
  }
}

.stepTitle {
  font-family: "Work Sans", sans-serif;
  font-weight: 600 !important;
  font-size: 1.4em;
  position: relative;
  margin-top: 20px;
  color: #d71920;

  @media only screen and (min-width: 1000px) {
    &[data-key="3"] {
      font-size: 2.6em;
    }
  }
}

.stepSubtitle {
  color: #777;
  font-family: "Work Sans", Arial, sans-serif;
  font-size: 15px;
  line-height: 23px;
  margin-top: 10px;
  max-width: 350px;
}

// ----------

.flipCard {
  background-color: transparent;
  width: 145px;
  height: 145px;
  perspective: 1000px;
  margin: 7px 8px;

  @media only screen and (min-width: 1000px) {
    &[data-key="3"] {
      width: 350px;
      height: 350px;
    }
  }
}

.flipCardInner {
  position: relative;
  width: 100%;
  height: 100%;
  text-align: center;
  transition: transform 0.6s;
  transform-style: preserve-3d;
}

.flipNow .flipCardInner {
  transform: rotateY(180deg);
}

.flipCardFront,
.flipCardBack {
  position: absolute;
  width: 100%;
  height: 100%;
  backface-visibility: hidden;
}

.flipCardBack {
  transform: rotateY(180deg);
}

.termsAndConditions {
  margin-top: 80px;
  line-height: 1.8rem;
  font-size: 1.2rem;
  letter-spacing: -0.05em;
  text-align: center;
  font-weight: 300;
  color: #777;
  padding-inline: 15px;

  a {
    text-decoration: none;
    color: #d71920;
    cursor: pointer;
  }
}