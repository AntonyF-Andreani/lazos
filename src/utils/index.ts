import { ImageLoaderProps } from "next/image";

/**
 * Utilidad para extraerla URL con formatos optimizados para las imagenes del CMS (Servicio de Imagenes)
 * https://www.storyblok.com/docs/image-service
 */
export const StoryBlockImageURL = ({
  src,
  width,
  quality = 20,
  format = "webp",
  height = 0,
}: StoryBlockImageURLParams) => {
  return `${src}/m/${width}x${height}/filters:quality(${quality}):format(${format})`;
};

interface StoryBlockImageURLParams {
  /** URL de la imagen como la provee Storyblok */
  src: string;
  /** Width of image. The Storyblok API infers aspect ratio */
  width: number;
  /** Quality of image, for default its 20 (20%). */
  quality?: number;
  /** Format of image, for default its webp */
  format?: "webp" | "jpeg" | "png";
  /** Height of the image */
  height?: number;
}

export const StoryBlockLoader = ({ src, width, quality = 20 }: ImageLoaderProps) => {
  return StoryBlockImageURL({ src, width, quality });
};

/**
 * Util for join classes undefined or string and filter undefined
 * @param args classNames
 *
 *  @example const className = joinClasses("a", "b", undefined, "c") // "a b c"
 */
export const joinClasses = (...args: (string | undefined)[]): string => {
  return args.filter((x) => x).join(" "); // undefined is falsey
};

/**
 * Util to test diferrent viewports, must be used with jest.
 *
 * This util make a mock implementation of window.matchMedia
 *
 * @example prepareViewport("(max-width: 768px)")
 */
export const prepareViewport = (mediaQuery: string) => {
  Object.defineProperty(window, "watchMedia", {
    writable: true,
    configurable: true,
    value: mediaQuery,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    return {
      matches: query === mediaQuery,
      media: "",
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};
