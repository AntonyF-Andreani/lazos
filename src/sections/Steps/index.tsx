import { Grid, Typography } from "@mui/material";
import React, { useEffect, useState } from "react";
import Image from "next/image";

import styles from "./index.module.scss";

import { joinClasses } from "utils";
import { Step, File } from "services/Cms/types";

export interface StepsProps {
  title: string;
  data: Step[];
  terms: File;
}

const Steps = ({ data, title, terms }: StepsProps) => {
  const [flip, setFlip] = useState(false);

  useEffect(() => {
    const id = setInterval(() => setFlip((prevFlip) => !prevFlip), 5000);

    return () => clearInterval(id);
  }, []);

  return (
    <div className={styles.section}>
      <Grid className={styles.container}>
        <Typography className={styles.sectionTitle}>{title}</Typography>
        <div className={styles.stepsMiscellaneous} />
        <Grid className={styles.stepsContainer}>
          {data.map(({ title, description, image, second_image }, i) => (
            <Grid key={i} className={styles.step} data-key={i}>
              <div
                className={joinClasses(
                  styles.flipCard,
                  flip && i == 3 ? styles.flipNow : undefined
                )}
                data-key={i}
                data-testid={"flipCard" + i}
              >
                <div className={styles.flipCardInner}>
                  <div className={styles.flipCardFront}>
                    <Image
                      alt="Avatar"
                      className={styles.stepImage}
                      data-key={i}
                      layout="fill"
                      src={image.filename}
                    />
                  </div>
                  {second_image && second_image.filename ? (
                    <div className={styles.flipCardBack}>
                      <Image
                        alt="Avatar-second"
                        className={styles.stepImage}
                        data-key={i}
                        layout="fill"
                        src={second_image.filename}
                      />
                    </div>
                  ) : null}
                </div>
              </div>

              <Typography className={styles.stepTitle} data-key={i}>
                {title}
              </Typography>
              <Typography className={styles.stepSubtitle}>{description}</Typography>
            </Grid>
          ))}
        </Grid>
        <Grid>
          <Typography className={styles.termsAndConditions}>
            Conocé los{" "}
            <a href={terms.filename} rel="noreferrer" target="_blank">
              términos y condiciones
            </a>{" "}
            para acceder a la Plataforma de Gestión Logística de Donaciones.
          </Typography>
        </Grid>
      </Grid>
    </div>
  );
};

export default Steps;
