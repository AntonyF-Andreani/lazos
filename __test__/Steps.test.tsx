import React from "react";

import { act, render } from "./utils";

import Steps, { StepsProps } from "sections/Steps";

const defaultProps: StepsProps = {
  title: "test",
  data: [
    {
      title: "title-test",
      description: "description-test",
      image: {
        filename: "/imagetest.jpg",
      },
      second_image: {
        filename: "/secondimagetest.jpg",
      },
    },
  ],
  terms: {
    filename: "/test.pdf",
  },
};

describe("<Steps />", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });
  it("should render", () => {
    const { asFragment } = render(<Steps {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render second image", () => {
    const { getByAltText } = render(<Steps {...defaultProps} />);

    const secondImage = getByAltText("Avatar-second");

    expect(secondImage).toBeInTheDocument();
  });
  it("should't render secondImage if is not present", () => {
    const { getAllByRole } = render(
      <Steps
        data={defaultProps.data.map((item) => ({ ...item, second_image: undefined }))}
        terms={defaultProps.terms}
        title={defaultProps.title}
      />
    );

    const onlyFirstImage = getAllByRole("img");

    expect(onlyFirstImage).toHaveLength(1);
  });

  it("should have terms in href", () => {
    const { getByRole } = render(<Steps {...defaultProps} />);

    const terms = getByRole("link");

    expect(terms.hasAttribute("href")).toBeTruthy();
    expect(terms.getAttribute("href")).toBe(defaultProps.terms.filename);
  });

  it("should render with specific class", async () => {
    const { getByTestId } = render(
      <Steps
        data={[
          ...defaultProps.data,
          ...defaultProps.data,
          ...defaultProps.data,
          ...defaultProps.data,
        ]}
        terms={defaultProps.terms}
        title={defaultProps.title}
      />
    );

    act(() => {
      jest.advanceTimersByTime(5000);
      jest.clearAllTimers();
    });

    expect(getByTestId("flipCard3")).toHaveClass("flipNow");
  });
});
