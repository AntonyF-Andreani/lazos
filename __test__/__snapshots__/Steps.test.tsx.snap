// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Steps /> should render 1`] = `
<DocumentFragment>
  <div
    class="section"
  >
    <div
      class="MuiGrid-root container css-vj1n65-MuiGrid-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 sectionTitle css-12hl0xw-MuiTypography-root"
      >
        test
      </p>
      <div
        class="stepsMiscellaneous"
      />
      <div
        class="MuiGrid-root stepsContainer css-vj1n65-MuiGrid-root"
      >
        <div
          class="MuiGrid-root step css-vj1n65-MuiGrid-root"
          data-key="0"
        >
          <div
            class="flipCard"
            data-key="0"
            data-testid="flipCard0"
          >
            <div
              class="flipCardInner"
            >
              <div
                class="flipCardFront"
              >
                <div
                  style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
                >
                  <img
                    alt="Avatar"
                    class="stepImage"
                    data-key="0"
                    data-nimg="fill"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </div>
              </div>
              <div
                class="flipCardBack"
              >
                <div
                  style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
                >
                  <img
                    alt="Avatar-second"
                    class="stepImage"
                    data-key="0"
                    data-nimg="fill"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </div>
              </div>
            </div>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 stepTitle css-12hl0xw-MuiTypography-root"
            data-key="0"
          >
            title-test
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 stepSubtitle css-12hl0xw-MuiTypography-root"
          >
            description-test
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root css-vj1n65-MuiGrid-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 termsAndConditions css-12hl0xw-MuiTypography-root"
        >
          Conocé los 
          <a
            href="/test.pdf"
            rel="noreferrer"
            target="_blank"
          >
            términos y condiciones
          </a>
           para acceder a la Plataforma de Gestión Logística de Donaciones.
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
